### ISSUE TITLE
lutimes for symlink file failed with "Stale file handle"

### ISSUE THREAD

**Issue by derfian** [unfs3/unfs3#4]
This issue was initially reported on SourceForge
_Link:_ https://sourceforge.net/p/unfs3/bugs/12/
_Reporter:_ Rui Wang
_Created:_ 2018-06-13

---
I am not sure if this is a bug. I found if I try to modify the timestamp of a symlink file pointing to a none existence file with lutimes(), the program will get an error: Stale file handle. But if the symlink file points to an exist file, the operation will success. I found this in yocto nfsroot. And I write a simple program to reproduce the issue.

```c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <errno.h>
#include <sys/time.h>

int main()
{
        int rc; 
        struct timeval stamps[2] = { 
                {.tv_sec = 0, .tv_usec = 0}, 
                {.tv_sec = 0, .tv_usec = 0}, 
        };

        printf("create symlink file test -> target\n");
        rc = symlink("target", "test");
        if (rc)
                perror("symlink");

        printf("**** target file is not exist ****\n");
        rc = lutimes("test", stamps);
        if (rc)
                perror("lutimes test");
        else
                printf("lutimes test success\n");

        system("touch target");

        printf("**** target file is exist ****\n");
        rc = lutimes("test", stamps);
        if (rc)
                perror("lutimes test");
        else
                printf("lutimes test success\n");

        system("rm test target");
        return 0;
}
```

**Comment by derfian** [comment]
_Commenter:_ @derfian 
_When:_ 2018-06-28

---
Sorry for not getting back to you earlier - I've been swamped with other things at work. I just wanted to say that I haven't forgotten about this yet. The bad news is that I'll be away on vacation for almost a month, but I hope to get back to this when I return.

**Comment by jschwartzenberg** [comment]
I'm getting `Stale NFS file handle` messages on the client as well from time to time on a reexported SMB mount. Unmounting and remounting solves it and I also just found that the problem can go away by itself after a while.

**Comment by CendioOssman** [comment]
That sounds like something else as SMB does not have symlinks. Please open a new issue and try to fill out as many details as you can.

**Comment by jschwartzenberg** [comment]
I already figured out the cause of the problem. The SMB mount is set up through autofs and autofs's mounting/unmounting is creating problems. I set the autofs timeout to 2 days now so that it doesn't unmount so fast. If you want I'll create another issue for this.

**Comment by ChenQi1989** [comment]
@derfian I rebased the codes against master and sent out a PR: https://github.com/unfs3/unfs3/pull/35
The problem could be reproduced by running `ln -s dest src && touch -h src'.
Apart from the rpm issue it mentions in the commit, docker pull also fails with this stale file error. There seems to be a common pattern here: extract files from some bundle and set times for them (using utimensat).

**Comment by CendioOssman** [comment]
This has been resolved in 1850aa8 now, based on #35.
